docs: Create the design section for the public dashboard feature (#1540)
Flank/flank-dashboard: metrics/web/docs/features/public_dashboard/01_public_dashboard.md

#### UserProfileViewModel
> Explain the purpose and responsibility of the class.

Since we want to disable some features on the UI if the user is anonymous, we should be able to access the user profile from the UI. As the [Presentation Layer Architecture](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/02_presentation_layer_architecture.md#view-model) document states, we should create a view model to be able to access the data from the `state` on the UI. So, we should create a view model representing a user profile to disable the `Project Groups` page and provide the ability to `Sign In` to the application for the anonymous users.

#### SignInAnonymouslyUseCase
> Explain the purpose and responsibility of the class.

Since the public dashboard feature implies logging in as an anonymous user, we should create a use case needed to log in to the application using the anonymous user. 

#### AuthState
> Explain the purpose and responsibility of the class.

> Since we should know whether the user signed in anonymously or using sign in providers or credentials, or logged out, we should provide an enum, which should contain all the described states. 

Reviewer: :pencil:
```suggestion
Since we should know the authentication state of the user, we should provide an enum that will contain the following states: 
 - authenticated
 - anonymous
 - notAuthenticated
```

### User Interface
> How users will interact with the feature (API, CLI, Graphical interface, etc.).

> Since the non-authorized users should not be able to access the `Project Groups` page, it should be hidden from the user menu for such users. Also, there should be an ability to sign in for anonymous users, so we should change the `Logout` button from the user menu to `Sign in` one. 

Reviewer: :readable:
```suggestion
As we've mentioned above, the non-authorized users should not be able to access the `Project Groups` page. Because of this, we should hide this option from the user menu for anonymous users. Also, there should be an ability to sign in for anonymous users, so we should change the `Logout` button from the user menu to the `Sign in` one. 
```

- The user logged in anonymously should be able to read the `project_groups` collection if the public dashboard feature is enabled.
- The anonymous user should not be able to write to the `project_groups` collection.
- The user logged in anonymously should be able to read only their profile document (document with the user identifier as document identifier) from the `user_profiles` collection if the public dashboard feature is enabled.
- The user logged in anonymously should be able to create/update their profile (document with the user identifier as document identifier) in the `user_profiles` collection if the public dashboard feature is enabled.
- The anonymous user should be able to read the data from the `build_days` collection if the public dashboard feature is enabled.
- The anonymous user should not be able to write to the `build_days` collection.

Please, note that if the collection is not mentioned in the list above, the collection rules should stay as is.

### Program
> Detailed solution description to class/method level.
 
The implementation of the public dashboard feature involves the changes in the following modules:

> - [the `feature config` module;](#featureconfig-module)
> - [the `auth` module;](#auth-module)
> - [the `navigation` module;](#navigation-module)
> 

Reviewer: :pencil:
```suggestion
- [The `feature config` module](#featureconfig-module).
- [The `auth` module](#auth-module).
- [The `navigation` module](#navigation-module).
```
Please, make lists across the document consistent.

Consider the following subsections that describe each module changes in more detail.

#### FeatureConfig module

The first important part is adding the ability to track the feature state, so that we can change the behavior of our application correspondingly.

To implement this, we should follow the next steps:
- add the `isPublicDashboardFeatureEnabled` boolean field to the `FeatureConfig` model;
- change the `FeatureConfigData`'s `fromJson()` and `toJson()` methods, so they include handling the new added field;
- create the [`PublicDashboardFeatureConfigViewModel`](#PublicDashboardFeatureConfigViewModel) and initialize it in the `FeatureConfigNotifier` as the other ones.

Let's review the following class diagram, which describes the relationship between the classes in this module:

![Feature Config Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_feature_config_class_diagram.puml)

#### Auth module

> The next one, we should add the ability to sign-in anonymously. Also, we should know whether the user sign-in anonymously. For these purposes, we should make the following changes in the `auth` module:

Reviewer 2: ```suggestion
Next, we should add the ability to sign in anonymously and detect this kind of user. For these purposes, we should make the following changes in the `auth` module:
```
What do you think?

- add the `signInAnonymously()` method to the `UserRepository` abstract class;
- add the implementation of the `signInAnonymously()` method to the `FirebaseUserRepository` class;
- create the [`SignInAnonymouslyUseCase`](#SignInAnonymouslyUseCase) class, which calls the `signInAnonymously()` method of the `UserRepository`;
- create the [`UserProfileViewModel`](#UserProfileViewModel) class;
- create the `AuthState` enum;
- integrate the use case, the enum, and the view model described above into the `AuthNotifier`.

Let's review the following class diagram, which describes the relationship between the classes in this module:

![Auth Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_auth_class_diagram.puml)

#### Navigation module

The last thing, we should have the ability to manage the navigation of the anonymous user, that is, to know which pages they can visit and which they can't. Following the next steps, helps to achieve this purpose:
- add the field of the `AuthState` type to the `NavigationNotifier`;

> - change the `handleAuthenticationUpdates()` method of the `NavigationNotifier` to take an argument of the type `AuthState`;

Reviewer 2: Take into account that according to this [document](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/features/deep_links/01_deep_linking.md#deep-linking-with-authorization), the "handleAuthenticationUpdates()" method will change, so I think we should use updated information.

- create a set of the pages' `RouteName`, which the anonymous user can visit (at the moment, it is only the `dashboard` route name);
- add the additional condition to the `_processConfiguration()` method, which should state the following, if the user auth state is logged in anonymously, and they visit the page, whose `RouteName` is not in the anonymous `RouteName` set, then redirect to the `login` page;

> #### Making things work

Reviewer: :readable:
I believe it would be great to have a sequence diagram in this section to explain how does the anonymous sign-in will work. Also, It would be great to have an overall class diagram to show how do the different models work with each other.

After the changes in the modules above, let's move on to integrating these changes into the application.

Here are the main parts that we should add to the application:
- auto sign in the user anonymously if the feature is enabled;
- manage normal sign in for the anonymous user;
- change the UI of the Metrics user menu.

Let's review each of the above points in the following subsections

##### Auto sign in the user anonymously

> To auto sign in the user anonymously, we should make the following:

Reviewer 2: Consider using a numbered list, according to the [Technical Writing](https://developers.google.com/tech-writing/one/lists-and-tables#choose_the_correct_type_of_list) article suggestions.

- subscribe the `NavigationNotifier` to the `AuthNotifier` so the `NavigationNotifier` is able to take the current user auth state from the `AuthNotifier` which in turn allows the `NavigationNotifier` to navigate depending on this value;

Consider the following component diagram that briefly describes this architecture approach:

![Public Dashboard Architecture Components Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_architecture_components_diagram.puml)

### User Interface
> How users will interact with the feature (API, CLI, Graphical interface, etc.).

As we've mentioned above, the non-authorized users should not be able to access the `Project Groups` page. Because of this, we should hide this option from the user menu for anonymous users. Also, there should be an ability to sign in for anonymous users, so we should change the `Logout` button from the user menu to the `Sign in` one. 

Let's consider the user menu UI in case we are using an anonymous user: 

> ![Anonymous User Menu](images/menu.png)

Reviewer 3: Hello👋. Can you check this point one more time? It looks like the picture is displayed fine in my browser.
![image](https://user-images.githubusercontent.com/87802654/128395133-701fa048-5120-4ca0-bdc1-7b819fac0720.png)

##### PublicDashboardFeatureConfigModel
> Explain the purpose and responsibility of the class.

Since we are going to store the public dashboard configuration in the Firestore database and the `FeatureConfigNotifier` manages the application state related to the feature configuration, we should be able to transfer the configuration of the public dashboard feature between different states. According to the [Presentation Layer Architecture](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/02_presentation_layer_architecture.md#model) document, to transfer the data between the states, we should use the `model` classes. So, for this purpose, we should create a `PublicDashboardFeatureConfigModel` that will be used to transfer the public dashboard feature configuration from the `FeatureConfigNotifier` to any other state in the application.

##### FeatureConfig module changes

> The list of changes this module requires to integrate the public dashboard feature:

Reviewer: :readable:
```suggestion
Consider the list of changes this module requires to integrate the public dashboard feature:
```

- add the `isPublicDashboardFeatureEnabled` boolean field to the `FeatureConfig` model; 
- change the `FeatureConfigData`'s `fromJson()` and `toJson()` methods, so they include handling the new added field;
- create the [`PublicDashboardFeatureConfigModel`](#PublicDashboardFeatureConfigModel) and initialize it in the `FeatureConfigNotifier`.

Let's review the following class diagram, which describes the relationship between the classes in this module:

![Feature Config Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_feature_config_class_diagram.puml)

#### Auth module

> Next, we should add the ability to sign in anonymously and detect this kind of user. For these purposes, we should create the classes and make module changes described in the following subsections.

Reviewer: 🏷️ 
```suggestion
Next, we should add the ability to sign in anonymously and detect the user type. Review sections below that describe changes needed to implement this functionality.
```
What do you think?

##### UserProfileViewModel
> Explain the purpose and responsibility of the class.

> Since we want to disable some features on the UI if the user is anonymous, we should be able to access the user profile from the UI. As the [Presentation Layer Architecture](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/02_presentation_layer_architecture.md#view-model) document states, we should create a view model to be able to access the data from the `state` on the UI. So, we should create a view model representing a user profile to disable the `Project Groups` page and provide the ability to `Sign In` to the application for the anonymous users.

Reviewer: 🏷️ 
```suggestion
Since we want to disable some features on the UI when the user is anonymous, we should be able to access the user profile from the UI. As the [Presentation Layer Architecture](https://github.com/Flank/flank-dashboard/blob/master/metrics/web/docs/02_presentation_layer_architecture.md#view-model) document states, we should create a `view model` to be able to access the data from the `state` on the UI. So, we should create a view model representing a user profile to disable the `Project Groups` page and provide the ability to `Sign In` to the application for the anonymous users.
```
?

##### SignInAnonymouslyUseCase
> Explain the purpose and responsibility of the class.

Since the public dashboard feature implies logging in as an anonymous user, we should create a use case needed to log in to the application using the anonymous user.

##### AuthState
> Explain the purpose and responsibility of the class.

Since we should know the authentication state of the user, we should provide an enum that will contain all the following states:
- loggedIn
- loggedInAnonymously
- loggedOut

##### Auth module changes

> The list of changes this module requires to integrate the public dashboard feature:

Reviewer: :readable:
```suggestion
Let's consider the list of changes this module requires to integrate the public dashboard feature:
```
?

- add the `signInAnonymously()` method to the `UserRepository` abstract class;
- add the implementation of the `signInAnonymously()` method to the `FirebaseUserRepository` class;
- add the `isPublicDashboardFeatureEnabled` field of the boolean type to the `AuthNotifier`;

> - add the `isAuthInitialized` field of the boolean type to the `AuthNotifier`;  

Reviewer: 🔖 
```suggestion
- add the `isInitialized` getter of the boolean type to the `AuthNotifier`;  
```
Does it make sense?

- create the [`UserProfileViewModel`](#UserProfileViewModel) class;
- create the [`SignInAnonymouslyUseCase`](#SignInAnonymouslyUseCase) class, which calls the `signInAnonymously()` method of the `UserRepository`;
- create the [`AuthState`](#AuthState) enum;
- integrate the view model, the use case, and the enum described above into the `AuthNotifier`.

Let's review the following class diagram, which describes the relationship between the classes in this module:

![Auth Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_auth_class_diagram.puml)

#### Navigation module

The last thing, we should have the ability to manage the navigation of the anonymous user, that is, to know which pages they can visit and which they can't. The next subsection describes changes that we should implement to integrate the public dashboard feature.

##### Navigation module changes

The list of changes this module requires to integrate the public dashboard feature:

> - add the field of the `AuthState` type to the `NavigationNotifier`;

Reviewer: 🔨 
Would it be great to mention the name of this field here?

- add the `handleAuthUpdates()` method to the `NavigationNotifier`, which should handle the updating of the `AuthState` in the `AuthNotifier`;

> - create a set of the pages' `RouteName`, which the anonymous user can visit (at the moment, it is only the `dashboard` route name);

Reviewer: 📱 
What do you think about adding the `bool` field into the `RouteConfiguration` class instead? Would it be a bit more clear and simple solution?

- add the additional condition to the `_processConfiguration()` method, which should state the following, if the user auth state is logged in anonymously, and they visit the page, whose `RouteName` is not in the anonymous `RouteName` set, then redirect to the `login` page;

#### Making things work

After the changes in the modules above, let's move on to integrating these changes into the application.

Here are the main parts that we should add to the application:
- auto sign in the user anonymously if the feature is enabled;
- manage normal sign in for the anonymous user;
- change the UI of the Metrics user menu.

Let's review each of the above points in the following subsections

##### Auto sign in the user anonymously

To auto sign in the user anonymously, we should make the following:
1. Find the `InjectionContainer`. 
2. Add the `handleAuthUpdates()` method of the `NavigationNotifier` to the listener of the `AuthNotifier`.
3. Create the listener for the `FeatureConfigNotifier` and add the `handlePublicDashboardFeatureConfigUpdates()` method of the `AuthNotifier`.

> 4. Add the logic of the signing the user in anonymously depending on the public dashboard feature and the auth states to the `AuthNotifier`.

Reviewer: :readable:
I believe it would be great to describe these steps in a bit more abstract way. More precisely, we should not describe the whole algorithm here, we should provide key points of this algorithm. For example: 
- The `handleAuthUpdates()` method of the `NavigationNotifier` should be called when the `AuthNotifier.authState` changes.

After this list of key points of this algorithm, we can explain the way of implementing some of them if there is any critical functionality/tricky moment in implementing them.

##### Manage sign in for the anonymous user

> Next one, to manage the sign in for the anonymous user, we should firstly sign out them. After, we should sign in the user normally using the sign in with credentials, or using the Google provider.

Reviewer: 🔨 
I think it would be great to take a look at the following situation and describe how we are going to deal with it. Imagine the anonymous user enters the login page and tries to log in. We have the following flow here: 
- Log out the current (anonymous) user. 
- Try to log in as a new user using the chosen approach. 

The question is: what are we going to do if the login process fails?

Consider the overall class diagram for the public dashboard feature that describes the relationships between the classes this feature requires:

![Public Dashboard Feature Class Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_class_diagram.puml)

Consider the overall sequence diagram for the public dashboard feature that describes the main flow for the public dashboard feature:

![Public Dashboard Feature Sequence Diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/Flank/flank-dashboard/public_dashboard_design/metrics/web/docs/features/public_dashboard/diagrams/public_dashboard_sequence_diagram.puml)

### Testing
> Describe an approach on how the feature should be tested.

To test the feature, we should focus on the following types of tests:
- The unit tests, which cover the main logic of the feature, described in the [program section](#program).

> - The widget tests, which test that the UI behaviour corresponds to the expectations from the [User Interface section](#user-interface).

Reviewer: 🔨
I believe we should also add integration tests to test that everything works correctly together.
